modules/core/src/main/scala/com.snowplowanalytics.iglu/client/validator/CirceValidator.scala

```diff
@@ -240,4 +246,65 @@ object CirceValidator extends Validator[Json] {
 
   private def fromValidationMessage(m: ValidationMessage): ValidatorReport =
     ValidatorReport(m.getMessage, m.getPath.some, m.getArguments.toList, m.getType.some)
+
+  type InitValidatorCache[F[_]] =
+    CreateLruMap[F, (SchemaKey, Int), Either[ValidatorError.InvalidSchema, JsonSchema]]
+  def validatorF[F[_]: Monad: InitValidatorCache]: F[ValidatorF[F, (SchemaKey, Int), Json]] =
+    CreateLruMap[F, (SchemaKey, Int), Either[ValidatorError.InvalidSchema, JsonSchema]]
+      .create(100)
+      .map(validatorFImpl(_))
+
+  def validatorFImpl[F[_]: Monad](
+    lookupJsonSchema: LruMap[F, (SchemaKey, Int), Either[ValidatorError.InvalidSchema, JsonSchema]]
+  ): ValidatorF[F, (SchemaKey, Int), Json] =
+    new ValidatorF[F, (SchemaKey, Int), Json] {
+
+      def lookupOrEvaluate(
+        key: Option[(SchemaKey, Int)],
+        schema: Json
+      ): F[Either[ValidatorError.InvalidSchema, JsonSchema]] = {
+        key
+          .fold(Monad[F].pure(none[Either[ValidatorError.InvalidSchema, JsonSchema]]))(
+            lookupJsonSchema.get(_)
+          )
+          .flatMap {
+            case Some(result) => result.pure[F]
+            case None =>
+              Either
+                .catchNonFatal(
+                  IgluMetaschemaFactory.getSchema(circeToJackson(schema), SchemaValidatorsConfig)
+                )
+                .leftMap(issue =>
+                  ValidatorError
+                    .InvalidSchema(
+                      NonEmptyList.of(ValidatorError.SchemaIssue("$", issue.getMessage))
+                    )
+                )
+                .pure[F]
+                .flatTap { result =>
+                  key.map(lookupJsonSchema.put(_, result)).sequence_
+                }
+          }
+      }
+
+      def validateF(
+        key: Option[(SchemaKey, Int)],
```

Comment:
How about we use here `Cached | NotCached` model, so we get rid of the Option? I think then we wouldn't need `key` arg, only `data` and Cached | NotCached `schema`, which would contain timestamp and schemaKey?